File: python/core/interfaces/ILLMProvider.py
"""
ILLMProvider - Python Interface for Language Model Providers
Enables switching between Azure OpenAI, Claude, local models, etc.
"""

from abc import ABC, abstractmethod
from typing import Dict, List, Optional, Any
from dataclasses import dataclass

@dataclass
class LLMResponse:
    content: str
    usage: Dict[str, int]
    model: str
    provider: str
    finish_reason: str = "stop"
    timestamp: str = None

@dataclass
class LLMMessage:
    role: str
    content: str
    name: Optional[str] = None

class ILLMProvider(ABC):
    """Abstract interface for Language Model providers"""
    
    @abstractmethod
    async def initialize(self) -> bool:
        """Initialize the provider and test connection"""
        pass
    
    @abstractmethod
    async def is_ready(self) -> bool:
        """Check if provider is ready to handle requests"""
        pass
    
    @abstractmethod
    def get_provider_name(self) -> str:
        """Get the provider name (e.g., 'Azure OpenAI', 'Claude')"""
        pass
    
    @abstractmethod
    def get_model_name(self) -> str:
        """Get the current model name"""
        pass
    
    @abstractmethod
    async def generate_response(
        self, 
        messages: List[LLMMessage], 
        max_tokens: Optional[int] = None,
        temperature: Optional[float] = None,
        **kwargs
    ) -> LLMResponse:
        """Generate response from messages"""
        pass
    
    @abstractmethod
    async def generate_with_tools(
        self,
        messages: List[LLMMessage],
        tools: List[Dict] = None,
        **kwargs
    ) -> LLMResponse:
        """Generate response with tool calling support"""
        pass
    
    @abstractmethod
    def get_autogen_config(self) -> Dict[str, Any]:
        """Get configuration for AutoGen compatibility"""
        pass
    
    @abstractmethod
    async def estimate_tokens(self, text: str) -> int:
        """Estimate token count for text"""
        pass
    
    @abstractmethod
    async def calculate_cost(self, input_tokens: int, output_tokens: int) -> Dict:
        """Calculate cost for token usage"""
        pass
    
    # Optional methods with default implementations
    async def supports_vision(self) -> bool:
        """Check if provider supports vision/image analysis"""
        return False
    
    async def supports_tools(self) -> bool:
        """Check if provider supports function/tool calling"""
        return False
    
    async def supports_streaming(self) -> bool:
        """Check if provider supports streaming responses"""
        return False
    
    async def get_health_status(self) -> Dict:
        """Get provider health status"""
        return {
            "provider": self.get_provider_name(),
            "model": self.get_model_name(),
            "status": "unknown",
            "ready": await self.is_ready()
        }

File: nodejs/src/core/interfaces/ILLMProvider.js
/**
 * LLM Provider Interface
 * Abstract interface for Language Model providers
 * Enables switching between Azure OpenAI, Claude, Google Gemini, local models, etc.
 */

class ILLMProvider {
    // Configuration and Connection
    async initialize() {
        throw new Error('initialize() must be implemented');
    }

    async testConnection() {
        throw new Error('testConnection() must be implemented');
    }

    async isReady() {
        throw new Error('isReady() must be implemented');
    }

    // Model Information
    getModelName() {
        throw new Error('getModelName() must be implemented');
    }

    getProviderName() {
        throw new Error('getProviderName() must be implemented');
    }

    getMaxTokens() {
        throw new Error('getMaxTokens() must be implemented');
    }

    getContextWindow() {
        throw new Error('getContextWindow() must be implemented');
    }

    // Token Management
    async estimateTokens(text) {
        throw new Error('estimateTokens() must be implemented');
    }

    async calculateCost(inputTokens, outputTokens) {
        throw new Error('calculateCost() must be implemented');
    }

    // Core LLM Operations
    async generateResponse(messages, options = {}) {
        throw new Error('generateResponse() must be implemented');
    }

    async generateStreamResponse(messages, options = {}, onChunk) {
        throw new Error('generateStreamResponse() must be implemented');
    }

    // Specialized Operations
    async generateEmbedding(text) {
        throw new Error('generateEmbedding() must be implemented');
    }

    async classifyIntent(message, classes) {
        throw new Error('classifyIntent() must be implemented');
    }

    async summarizeConversation(messages, maxLength = 200) {
        throw new Error('summarizeConversation() must be implemented');
    }

    // Function/Tool Calling (for advanced models)
    async generateWithTools(messages, tools = [], options = {}) {
        throw new Error('generateWithTools() must be implemented');
    }

    supportsTools() {
        return false; // Override in implementations that support tools
    }

    // Vision Capabilities (for multimodal models)
    async analyzeImage(imageData, prompt = "Describe this image") {
        throw new Error('analyzeImage() must be implemented');
    }

    supportsVision() {
        return false; // Override in implementations that support vision
    }

    // Audio/Speech Capabilities
    async speechToText(audioData) {
        throw new Error('speechToText() must be implemented');
    }

    async textToSpeech(text, voice = 'alloy') {
        throw new Error('textToSpeech() must be implemented');
    }

    supportsSpeech() {
        return false; // Override in implementations that support speech
    }

    // AutoGen Configuration
    getAutoGenConfig() {
        throw new Error('getAutoGenConfig() must be implemented');
    }

    // Error Handling and Retry Logic
    async handleAPIError(error, retryCount = 0) {
        throw new Error('handleAPIError() must be implemented');
    }

    // Rate Limiting
    async checkRateLimit() {
        return true; // Override if rate limiting is needed
    }

    // Model Switching (for providers with multiple models)
    async switchModel(modelName) {
        throw new Error('switchModel() must be implemented');
    }

    getAvailableModels() {
        throw new Error('getAvailableModels() must be implemented');
    }

    // Health Monitoring
    getHealthStatus() {
        return {
            provider: this.getProviderName(),
            model: this.getModelName(),
            status: 'unknown',
            lastCheck: null,
            uptime: 0,
            errorCount: 0,
            totalRequests: 0
        };
    }

    async performHealthCheck() {
        throw new Error('performHealthCheck() must be implemented');
    }
}

module.exports = ILLMProvider;
